main.go

```diff
@@ -18,45 +22,130 @@ var posts = []post{{
 	Title: "YOusdkfkd",
 }}
 
-func postsHandler(w http.ResponseWriter, r *http.Request) {
-	w.Header().Set("Content-Type", "application/json")
-
-	handlers := map[string]http.HandlerFunc{
-		"GET": func(w http.ResponseWriter, r *http.Request) {
-			resp, _ := json.Marshal(posts)
-			fmt.Fprintf(w, string(resp))
-		},
-		"POST": func(w http.ResponseWriter, r *http.Request) {
-			decoder := json.NewDecoder(r.Body)
-			var p post
-			if err := decoder.Decode(&p); err != nil {
-				panic(err)
-			}
-
-			id := len(posts) + 1
-			p.ID = id
-			posts = append(posts, p)
-			resp, _ := json.Marshal(p)
-			fmt.Fprintf(w, string(resp))
-		},
-		"DELETE": func(w http.ResponseWriter, r *http.Request) {
-			println(r.URL.Path)
-		},
-	}
-
-	handler, ok := handlers[r.Method]
-	if !ok {
-		fmt.Fprintf(w, "Method not supported")
+func find(id int) (int, error) {
+	for i, v := range posts {
+		if v.ID == id {
+			return i, nil
+		}
+	}
+
+	return 0, errors.New("Not found")
+}
+
+func getPosts(w http.ResponseWriter, r *http.Request) {
+	resp, err := json.Marshal(posts)
+	if err != nil {
+		panic(err)
+	}
+
+	fmt.Fprintf(w, string(resp))
+}
+
+func getPost(w http.ResponseWriter, r *http.Request) {
+	id, err := strconv.Atoi(chi.URLParam(r, "id"))
+	if err != nil {
+		panic(err)
+	}
+
+	i, err := find(id)
+	if err != nil {
+		fmt.Fprintf(w, err.Error())
 		return
 	}
 
-	handler(w, r)
+	post := posts[i]
+	resp, err := json.Marshal(post)
+	if err != nil {
+		panic(err)
+	}
+
+	fmt.Fprintf(w, string(resp))
 }
 
-func main() {
-	http.HandleFunc("/posts", postsHandler)
+func updatePost(w http.ResponseWriter, r *http.Request) {
+	id, err := strconv.Atoi(chi.URLParam(r, "id"))
+	if err != nil {
+		panic(err)
+	}
+
+	decoder := json.NewDecoder(r.Body)
+	var p post
+	if err := decoder.Decode(&p); err != nil {
+		panic(err)
+	}
+
+	i, err := find(id)
+	if err != nil {
+		fmt.Fprintf(w, err.Error())
+		return
+	}
+
+	updated := posts[i]
+	updated.Body = p.Body
+	updated.Title = p.Title
+	posts[i] = updated
```

Comment:
Not that accessing `posts` anywhere inside handler code will potentially cause a race condition, since concurrent HTTP requests run concurrently. The easy way to fix this is to wrap this global variable in a mutex whenever it is accessed: https://golang.org/pkg/sync/#Mutex. You use `.Lock()` whenever you want to read or write to it, and `.Unlock` as soon as you're done. See https://tour.golang.org/concurrency/9 for more of an example of this pattern.

Generally speaking though, using global variables is often avoided in Go as much as possible. In your current set up, it isn't that simple to fix, but if you created a struct which then implemented these functions as methods on itself, you could attach both `posts` and the mutex to the struct, which would make it much nicer. We can talk a bit deeper about that if you want, let me know on slack.